dev/src/document-group.ts

```diff
@@ -0,0 +1,347 @@
+/*!
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+import deepEqual from 'deep-equal';
+import * as through2 from 'through2';
+
+import {google} from '../protos/firestore_proto_api';
+import api = google.firestore.v1beta1;
+
+import {DocumentSnapshot} from './document';
+import {logger} from './logger';
+import {FieldPath} from './path';
+import {DocumentReference, QuerySnapshot} from './reference';
+import {AnyDuringMigration} from './types';
+import {requestTag} from './util';
+import {Watch} from './watch';
+
+/**
+ * A `DocumentGroup` combines read operations for multiple documents and
+ * supports efficient retrieval for groups of documents from a Firestore
+ * database.
+ *
+ * @class DocumentGroup
+ */
+export class DocumentGroup {
+  private readonly validator: AnyDuringMigration;
+
+  /**
+   * @private
+   * @hideconstructor
+   *
+   * @param {Firestore} _firestore - The Firestore Database client.
+   * @param {DocumentReference[]} documents The documents to combine in this
+   * document group.
+   * @param {{}=} options - Read options such as transaction ID and field mask.
+   */
+  constructor(
+      private readonly _firestore: AnyDuringMigration,
+      private readonly documents: DocumentReference[],
+      private readonly options:
+          {transactionId?: Uint8Array, fieldMask?: FieldPath[]} = {}) {
+    this.validator = _firestore._validator;
+  }
+
+  /**
+   * The [Firestore]{@link Firestore} instance for the Firestore
+   * database (useful for performing transactions, etc.).
+   *
+   * @type {Firestore}
+   * @name Query#firestore
+   * @readonly
+   *
+   * @example
+   * let docA = firestore.doc('coll/a');
+   * let docB = firestore.doc('coll/b');
+   *
+   * let documentGroup = firestore.documentGroup(docA, docB);
+   *
+   * let firestore = documentGroup.firestore;
+   */
+  get firestore() {
```

Comment:
Note this is getter is only defined for its docstring. Otherwise I would have inlined it.